packages/@aws-cdk/aws-lambda-python/lib/bundling.ts

```diff
@@ -16,39 +26,106 @@ export interface BundlingOptions {
    * The runtime of the lambda function
    */
   readonly runtime: lambda.Runtime;
+
+  /**
+   * Commands to run
+   */
+  readonly depsCommand: string;
 }
 
 /**
  * Produce bundled Lambda asset code
  */
 export function bundle(options: BundlingOptions): lambda.AssetCode {
-  // Bundling image derived from runtime bundling image (AWS SAM docker image)
-  const image = cdk.BundlingDockerImage.fromAsset(__dirname, {
-    buildArgs: {
-      IMAGE: options.runtime.bundlingDockerImage.image,
+  const { entry, runtime, depsCommand } = options;
+
+  // Determine which bundling image to use.
+  let image: cdk.BundlingDockerImage;
+  if (hasDependencies(entry)) {
+    // When dependencies are present, we use a Dockerfile that can create a
+    // cacheable layer. We can't use this Dockerfile if there aren't
+    // dependencies or the Dockerfile will complain about missing sources.
+    image = cdk.BundlingDockerImage.fromAsset(entry, {
+      buildArgs: {
+        IMAGE: runtime.bundlingDockerImage.image,
+      },
+      file: path.join(__dirname, 'Dockerfile.dependencies'),
+    });
+  } else {
+    // When dependencies aren't present, we can use the default Dockerfile that
+    // doesn't need to create a cacheable layer.
+    image = cdk.BundlingDockerImage.fromAsset(entry, {
+      buildArgs: {
+        IMAGE: runtime.bundlingDockerImage.image,
+      },
+      file: path.join(__dirname, 'Dockerfile'),
+    });
+  }
+
+  return lambda.Code.fromAsset(entry, {
+    assetHashType: cdk.AssetHashType.BUNDLE,
+    exclude: DEPENDENCY_EXCLUDES,
+    bundling: {
+      image,
+      command: ['bash', '-c', depsCommand],
     },
   });
+}
 
-  let installer = options.runtime === lambda.Runtime.PYTHON_2_7 ? Installer.PIP : Installer.PIP3;
+export interface BundleFunctionOptions {
+  readonly entry: string;
+  readonly runtime: lambda.Runtime;
+}
 
-  let hasRequirements = fs.existsSync(path.join(options.entry, 'requirements.txt'));
+export function bundleFunction(options: BundleFunctionOptions): lambda.AssetCode {
+  const { entry, runtime } = options;
 
-  let depsCommand = chain([
-    hasRequirements ? `${installer} install -r requirements.txt -t ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR}` : '',
+  const depsCommand = chain([
+    hasDependencies(entry) ? `rsync -r ${BUNDLER_DEPENDENCIES_CACHE}/. ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR}` : '',
     `rsync -r . ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR}`,
   ]);
 
-  return lambda.Code.fromAsset(options.entry, {
-    bundling: {
-      image,
-      command: ['bash', '-c', depsCommand],
-    },
+  return bundle({
+    entry,
+    runtime,
+    depsCommand,
+  });
+}
+
+export interface BundleLayerOptions {
+  readonly entry: string;
+  readonly runtime: lambda.Runtime;
+}
+
+export function bundleLayer(options: BundleLayerOptions): lambda.AssetCode {
+  const { entry, runtime } = options;
```

Comment:
Can bundleLayer and bundle be combined or do you feel keeping them separate is warranted?

Reply:
I was on the fence about this already. I like your idea.